File: src/services/TheBrain/src/domain/geolocation/GeoLocationCalculator.ts
import IGeoLocationCalculator from "./IGeoLocationCalculator";

export default class GeoLocationCalculator implements IGeoLocationCalculator {
    private readonly EarthRadiusInMeters : number = 6371000;
    private readonly EarthRadiusInKilometers : number = 6371;

    isInsideRadius(
        centerLatitude: number, 
        centerLongitude: number, 
        targetLatitude: number, 
        targetLongitude: number,
        radius : number
        ): boolean {
        const distance = this.calculateDistance(
            centerLatitude, 
            centerLongitude, 
            targetLatitude, 
            targetLongitude
            );
        return distance <= radius;
    }

    calculateDistance(
        firstLatitude: number,
        firstLongitude: number, 
        secondLatitude: number, 
        secondLongitude: number
        ): number {
        const latitudeDifference = this.degreesToRadians(secondLatitude - firstLatitude);
        const longitudeDifference = this.degreesToRadians(secondLongitude - firstLongitude);
        
        const haversineCentralAngle =
            Math.sin(latitudeDifference / 2) * Math.sin(latitudeDifference / 2) +
            Math.cos(this.degreesToRadians(firstLatitude)) *
            Math.cos(this.degreesToRadians(secondLatitude)) *
            Math.sin(longitudeDifference / 2) *
            Math.sin(longitudeDifference / 2);
        
        const radiansCentralAngle = 2 * Math.atan2(Math.sqrt(haversineCentralAngle), Math.sqrt(1 - haversineCentralAngle));
        const distance = this.EarthRadiusInKilometers * radiansCentralAngle;
        return distance;
    }

    degreesToRadians(degrees: number): number {
        return degrees * (Math.PI / 180);
    }

    kilometerRadiusToDegreesRadius(kilometerRadius: number): number {
        return kilometerRadius / this.EarthRadiusInMeters * (180 / Math.PI);
    }
    
}

File: src/services/NationalGeographicMessager/Domain/GeolocationAggregated/GeoLocationCalculator.cs
namespace NationalGeographicMessager.Domain.GeolocationAggregated
{
    public class GeoLocationCalculator : IGeoLocationCalculator
    {
        private const double EarthRadiusInMeters = 6371000;
        private const int EarthRadiusInKilometers = 6371;

        public bool IsInsideRadius(
            double centerLatitude, 
            double centerLongitude, 
            double targetLatitude, 
            double targetLongitude, 
            double radius
            )
        {
            var distance = CalculateDistance(
                centerLatitude,
                centerLongitude,
                targetLatitude,
                targetLongitude
            );

            return distance <= radius;
        }

        public double KilometerRadiusToDegreesRadius(double kilometerRadius)
        {
            return kilometerRadius / EarthRadiusInMeters * (180 / Math.PI);
        }

        internal static double CalculateDistance(
            double firstLatitude,
            double firstLongitude,
            double secondLatitude,
            double secondLongitude
            )
        {
            var latitudeDifference = DegreesToRadians(secondLatitude - firstLatitude);
            var longitudeDifference = DegreesToRadians(secondLongitude - firstLongitude);

            var haversineCentralAngle =
                Math.Sin(latitudeDifference / 2) * Math.Sin(latitudeDifference / 2) +
                Math.Cos(DegreesToRadians(firstLatitude)) *
                Math.Cos(DegreesToRadians(secondLatitude)) *
                Math.Sin(longitudeDifference / 2) *
                Math.Sin(longitudeDifference / 2);

            var radiansCentralAngle = 2 * Math.Atan2(Math.Sqrt(haversineCentralAngle), Math.Sqrt(1 - haversineCentralAngle));
            var distance = EarthRadiusInKilometers * radiansCentralAngle;
            return distance;
        }

        internal static double DegreesToRadians(double degrees)
        {
            return degrees * (Math.PI / 180);
        }

    }
}
